dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp

```diff
@@ -481,33 +482,101 @@ SegmentPtr DeltaMergeStore::segmentMergeDelta(
+                return {};
+            }
+        }
+    }
+
+    Segment::IngestDataInfo ingest_info;
+    if (clear_all_data_in_segment)
+        ingest_info = segment->prepareIngestDataWithClearData();
+    else
+        ingest_info = segment->prepareIngestDataWithPreserveData(dm_context, snapshot);
+
+    SegmentPtr new_segment{};
     {
         std::unique_lock lock(read_write_mutex);
         if (!isSegmentValid(lock, segment))
         {
-            LOG_DEBUG(log, "ReplaceData - Give up segment replace data because segment not valid, segment={} data_file={}", segment->simpleInfo(), data_file->path());
+            LOG_DEBUG(log, "IngestData - Give up segmentIngestData because segment not valid, segment={}", segment->simpleInfo());
             return {};
         }
 
         auto segment_lock = segment->mustGetUpdateLock();
-        new_segment = segment->replaceData(segment_lock, dm_context, data_file);
+        // Note: applyIngestData itself writes the wbs, and we don't need to pass a wbs by ourselves.
+        auto apply_result = segment->applyIngestData(segment_lock, dm_context, data_file, ingest_info);
 
-        RUNTIME_CHECK(compare(segment->getRowKeyRange().getEnd(), new_segment->getRowKeyRange().getEnd()) == 0, segment->info(), new_segment->info());
-        RUNTIME_CHECK(segment->segmentId() == new_segment->segmentId(), segment->info(), new_segment->info());
-
-        segment->abandon(dm_context);
-        segments[segment->getRowKeyRange().getEnd()] = new_segment;
-        id_to_segment[segment->segmentId()] = new_segment;
+        if (auto * const result = std::get_if<Segment::IngestDataResultSegmentReplaced>(&apply_result); result)
+        {
+            new_segment = result->segment;
+
+            RUNTIME_CHECK(
+                compare(segment->getRowKeyRange().getEnd(), new_segment->getRowKeyRange().getEnd()) == 0,
+                segment->info(),
+                new_segment->info());
+            RUNTIME_CHECK(
+                segment->segmentId() == new_segment->segmentId(),
+                segment->info(),
+                new_segment->info());
+
+            segment->abandon(dm_context);
+            segments[segment->getRowKeyRange().getEnd()] = new_segment;
+            id_to_segment[segment->segmentId()] = new_segment;
+
+            LOG_INFO(
+                log,
+                "IngestData - Finish, new segment is created, old_segment={} new_segment={}",
+                segment->info(),
+                new_segment->info());
+        }
+        else if (std::holds_alternative<Segment::IngestDataResultError>(apply_result))
+        {
+            // This should not happen, because we have verified segment is not abandoned.
+            LOG_DEBUG(log, "IngestData - Give up segmentIngestData because applyIngestData failed (usually this should not happen), segment={}", segment->simpleInfo());
+            return {};
```

Comment:
We should throw an exception or at least use `LOG_ERROR` here to indicate something went wrong. Or it may run into an inf retrying loop without any clear error message.